Modules/Loadable/Markups/MRML/vtkMRMLMarkupsCurveNode.cxx

```diff
@@ -183,6 +194,182 @@ double vtkMRMLMarkupsCurveNode::GetCurveLengthBetweenStartEndPointsWorld(vtkIdTy
+	    {
+	    originalControlPoints->GetPoint(segmentStartIndex - 1, segmentEndPoint);
+	    }
+	  else
+		{
+		double* segmentEndPoint1 = originalControlPoints->GetPoint(segmentStartIndex - 1);
+		double dist1 = vtkMath::Distance2BetweenPoints(segmentEndPoint1, projectedPoint);
+		double* segmentEndPoint2 = originalControlPoints->GetPoint(segmentStartIndex + 1);
+		double dist2 = vtkMath::Distance2BetweenPoints(segmentEndPoint2, projectedPoint);
+		if ((dist1 < dist2) && dist1 < vtkMath::Distance2BetweenPoints(segmentEndPoint1, segmentStartPoint))
+		  {
+		  segmentEndPoint[0] = segmentEndPoint1[0];
+		  segmentEndPoint[1] = segmentEndPoint1[1];
+		  segmentEndPoint[2] = segmentEndPoint1[2];
+		  }
+		else
+		  {
+		  segmentEndPoint[0] = segmentEndPoint2[0];
+		  segmentEndPoint[1] = segmentEndPoint2[1];
+		  segmentEndPoint[2] = segmentEndPoint2[2];
+		  }
+		}
+	  // estimate normal vector from control points on either side of projected point
+	  vtkIdType pointIdStart = pointLocator->FindClosestPoint(segmentStartPoint);
+	  double startNormal[3] = { 0.0 };
+	  normalVectorArray->GetTuple(pointIdStart, startNormal);
+	  vtkIdType pointIdEnd = pointLocator->FindClosestPoint(segmentEndPoint);
+	  double endNormal[3] = { 0.0 };
+	  normalVectorArray->GetTuple(pointIdStart, endNormal);
+	  double rayDirection[3] = { 0.0 };
+	  rayDirection[0] = (startNormal[0] + endNormal[0]) / 2.0;
+	  rayDirection[1] = (startNormal[1] + endNormal[1]) / 2.0;
+	  rayDirection[2] = (startNormal[2] + endNormal[2]) / 2.0;
+	  pointNormalArray->InsertNextPoint(rayDirection);
+		
+	  }
+
+	vtkNew<vtkPoints> snappedToSurfaceControlPoints;
+	vtkMRMLMarkupsCurveNode::ConstrainPointsToSurface(interpolatedPoints, pointNormalArray, surfacePolydata, snappedToSurfaceControlPoints, maximumSearchRadius);
+
+	this->SetControlPointPositionsWorld(snappedToSurfaceControlPoints);
+	this->SetControlPointLabelsWorld(originalLabels, originalControlPoints);
+}
+//---------------------------------------------------------------------------
+bool vtkMRMLMarkupsCurveNode::ConstrainPointsToSurface(vtkPoints* originalPoints, vtkPoints* normalVectors, vtkPolyData* surfacePolydata,
+  vtkPoints* surfacePoints, double maximumSearchRadius)
+{
+  if (originalPoints->GetNumberOfPoints()!= normalVectors->GetNumberOfPoints())
+	{
+	vtkGenericWarningMacro("vtkMRMLMarkupsCurveNode::ResamplePoints failed: invalid inputs");
+	return false;
+	}
+	
+  vtkNew<vtkOBBTree> surfaceObbTree;
+  surfaceObbTree->SetDataSet(surfacePolydata);
+  surfaceObbTree->BuildLocator();
+  double tolerence = surfaceObbTree->GetTolerance();
+
+  vtkNew<vtkPointLocator> pointLocator;
+  pointLocator->SetDataSet(surfacePolydata);
+  pointLocator->BuildLocator();
+
+  double originalPoint[3] = { 0.0 };
+  double rayDirection[3] = { 0.0 };
+  double exteriorPoint[3] = { 0.0 };
+
+  // Curve is expected to be close to surface. If projection distance is greater than 10% of image size
```

Comment:
It seems that the 10% is not hardcoded anymore so this comment should be updated.